File: java/src/main/java/toolgood/algorithm/internals/functions/mathsum/Function_COUNTIF.java
package toolgood.algorithm.internals.functions.mathsum;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.FunctionUtil;
import toolgood.algorithm.internals.functions.Function_2;
import toolgood.algorithm.internals.functions.NoneEngine;
import toolgood.algorithm.internals.functions.Tuple;

public final class Function_COUNTIF extends Function_2 {
    public Function_COUNTIF(FunctionBase[] funcs) {
        super(funcs);
    }

    @Override
    public String Name() {
        return "CountIf";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, java.util.function.BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = GetArray_1(engine, tempParameter);
        if (args1.IsErrorOrNone()) {
            return args1;
        }

        Operand args2 = func2.Evaluate(engine, tempParameter);
        if (args2.IsErrorOrNone()) {
            return args2;
        }

        List<BigDecimal> list = new ArrayList<>();
        boolean o = FunctionUtil.FlattenToList_Operand_BigDecimal(args1, list);
        if (o == false) {
            return ParameterError(1);
        }
        int count;
        if (args2.IsNumber()) {
            count = FunctionUtil.GetCountIf(list, args2.NumberValue());
        } else {
            String text = args2.TextValue().trim();
            try {
                BigDecimal d = new BigDecimal(text);
                count = FunctionUtil.GetCountIf(list, d);
            } catch (NumberFormatException e) {
                Tuple<String, BigDecimal> m2 = FunctionUtil.ParseSumIfMatch(text);
                if (m2 != null) {
                    count = FunctionUtil.GetCountIf(list, m2.getItem1(), m2.getItem2());
                } else {
                    return ParameterError(2);
                }
            }
        }
        return Operand.Create(count);
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.ARRAY);
        func2.GetParameterTypes(noneEngine, result, OperandType.NONE);
    }
}

File: csharp/ToolGood.Algorithm/Internals/Functions/MathSum/Function_COUNTIF.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using ToolGood.Algorithm.Enums;
using ToolGood.Algorithm.Internals;

namespace ToolGood.Algorithm.Internals.Functions.MathSum
{
	internal sealed class Function_COUNTIF : Function_2
    {
		public Function_COUNTIF(FunctionBase[] funcs) : base(funcs)
		{
		}

        public override string Name => "CountIf";

        public override Operand Evaluate(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
        {
            var args1 = GetArray_1(engine, tempParameter);
            if (args1.IsErrorOrNone) { return args1; }

            var args2 = func2.Evaluate(engine, tempParameter);
            if (args2.IsErrorOrNone) { return args2; }

            var list = new List<decimal>();
            var o = FunctionUtil.FlattenToList(args1, list);
            if (o == false) { return ParameterError(1); }
            int count;
            if (args2.IsNumber) {
                count = FunctionUtil.GetCountIf(list, args2.NumberValue);
            } else {
                var span = args2.TextValue.AsSpan().Trim();
                if (decimal.TryParse(span, NumberStyles.Any, CultureInfo.InvariantCulture, out decimal d)) {
                    count = FunctionUtil.GetCountIf(list, d);
                } else {
                    var m2 = FunctionUtil.ParseSumIfMatch(args2.TextValue.Trim());
                    if (m2 != null) {
                        count = FunctionUtil.GetCountIf(list, m2.Item1, m2.Item2);
                    } else {
                        return ParameterError(2);
                    }
                }
            }
            return Operand.Create(count);
        }
		public override OperandType GetResultType()
		{
			return OperandType.NUMBER;
		}

		internal override void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, string op = null, string val = null)
		{
			func1.GetParameterTypes(noneEngine, result, OperandType.ARRAY);
			func2.GetParameterTypes(noneEngine, result, OperandType.NONE);
		}
	}

}
